assets/src/loader.rs

```diff
@@ -0,0 +1,180 @@
+              A::Context: Send + Sync + 'static,
+              A::Error: Send + 'static,
+              F: Format<Data = A::Data> + Send + 'static,
+              F::Error: Send + 'static,
+              N: Into<String>
+    {
+        self.load_from::<A, F, _, _>(id, format, self.directory.clone())
+    }
+
+    /// Loads an asset with a given id and format from a custom store.
+    /// The actual work is done on a worker thread, thus this method immediately returns
+    /// a future.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the asset wasn't registered.
+    pub fn load_from<A, F, N, S>(&self,
+                                 name: N,
+                                 format: F,
+                                 store: Arc<S>)
+                                 -> AssetFuture<A, AssetError<A::Error, F::Error, S::Error>>
+        where A: Asset + Send + 'static,
+              A::Context: Send + Sync + 'static,
+              A::Error: Send + 'static,
+              F: Format<Data = A::Data> + Send + 'static,
+              F::Error: Send + 'static,
+              N: Into<String>,
+              S: Store + Send + Sync + 'static,
+              S::Error: Send
+    {
+        let context = self.contexts
+            .get(&TypeId::of::<A>())
+            .expect("Assets need to be registered");
+
+        let context: &Arc<A::Context> = context.downcast_ref().unwrap();
+
+        load_asset_future::<A, F, N, S>(context.clone(), format, name, store, &*self.pool)
+    }
+}
+
+/// Loads an asset with a given context, format, specifier and storage right now.
+pub fn load_asset<A, F, N, S>(context: &A::Context,
+                              format: &F,
+                              name: N,
+                              storage: &S)
+                              -> Result<A, AssetError<A::Error, F::Error, S::Error>>
+    where A: Asset,
+          F: Format<Data = A::Data>,
+          N: Into<String>,
+          S: Store
+{
+    let name = name.into();
+
+    let store_id = storage.store_id();
+    let spec = AssetSpec::new(name.clone(), F::extension(), store_id);
+
+    A::cached(context, &spec)
+        .map(|a| Ok(a))
+        .unwrap_or_else(move || load_asset_inner(context, format, spec, storage))
+        .map_err(|e| AssetError::new(AssetSpec::new(name, F::extension(), store_id), e))
+}
+
+fn load_asset_inner<A, F, S>(context: &A::Context,
+                             format: &F,
+                             spec: AssetSpec,
+                             storage: &S)
+                             -> Result<A, LoadError<A::Error, F::Error, S::Error>>
+    where A: Asset,
+          F: Format<Data = A::Data>,
+          S: Store
+{
+    let bytes = storage
+        .load(A::category(), &spec.name, spec.ext)
+        .map_err(LoadError::StorageError)?;
+    let data = format.import(bytes).map_err(LoadError::FormatError)?;
+    let a = Asset::from_data(data, context)
+        .map_err(LoadError::AssetError)?;
```

Comment:
That's not possible; I didn't define the error type I'm getting an instance of here, it's an associated type implementing `Error`.